conjure-error/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+
+extern crate self as conjure_error;
+
+use conjure_object::{Uuid, Value};
+use serde::Serialize;
+
+use crate::ser::ParametersSerializer;
+
+mod ser;
+#[allow(clippy::all)]
+mod types;
+
+pub use crate::types::*;
+
+impl ErrorCode {
+    /// Returns the HTTP status code associated with the error code.
+    #[inline]
+    pub fn status_code(&self) -> u16 {
+        match self {
+            ErrorCode::PermissionDenied => 403,
+            ErrorCode::InvalidArgument => 400,
+            ErrorCode::NotFound => 404,
+            ErrorCode::Conflict => 409,
+            ErrorCode::RequestEntityTooLarge => 413,
+            ErrorCode::FailedPrecondition => 500,
+            ErrorCode::Internal => 500,
+            ErrorCode::Timeout => 500,
+            ErrorCode::CustomClient => 400,
+            ErrorCode::CustomServer => 500,
+        }
+    }
+}
+
+/// A trait implemented by Conjure error types.
+pub trait ErrorType {
+    /// Returns the error's code.
+    fn code(&self) -> ErrorCode;
+
+    /// Returns the error's name.
+    ///
+    /// The name must be formatted like `NamespaceName:ErrorName`.
+    fn name(&self) -> &str;
+
+    /// Returns a sorted slice of the names of the error's safe parameters.
+    fn safe_args(&self) -> &'static [&'static str];
+}
+
+/// Encodes a Conjure error into its serialized form.
+///
+/// The error's instance ID will be randomly generated.
+///
+/// # Panics
+///
+/// Panics if the error type does not serialize as a struct.
+pub fn encode<T>(error: &T) -> SerializableError
+where
+    T: ErrorType + Serialize,
+{
+    let mut builder = SerializableError::builder();
+
+    let parameters = error
+        .serialize(ParametersSerializer)
+        .expect("failed to serialize error parameters");
+
+    for (key, value) in parameters {
+        if let Some(value) = value_string(&value) {
+            builder.insert_parameters(key, value);
+        }
+    }
+
+    builder
+        .error_code(error.code())
+        .error_name(error.name())
+        .error_instance_id(Uuid::new_v4())
+        .build()
+}
+
+fn value_string(value: &Value) -> Option<String> {
+    match value {
+        Value::Bool(v) => Some(v.to_string()),
+        Value::U8(v) => Some(v.to_string()),
+        Value::U16(v) => Some(v.to_string()),
+        Value::U32(v) => Some(v.to_string()),
+        Value::U64(v) => Some(v.to_string()),
+        Value::I8(v) => Some(v.to_string()),
+        Value::I16(v) => Some(v.to_string()),
+        Value::I32(v) => Some(v.to_string()),
+        Value::I64(v) => Some(v.to_string()),
+        Value::F32(v) => Some(v.to_string()),
+        Value::F64(v) => Some(v.to_string()),
+        Value::Char(v) => Some(v.to_string()),
+        Value::String(v) => Some(v.to_string()),
+        Value::Unit => None,
+        Value::Option(Some(v)) => value_string(v),
+        Value::Option(None) => None,
+        Value::Newtype(v) => value_string(v),
+        Value::Seq(_) | Value::Map(_) | Value::Bytes(_) => None,
```

Comment:
This is different behavior than the Java implementation, but this is a reasonable compromise given that the Java implementation doesn't match the spec.